common4j/src/main/com/microsoft/identity/common/java/cache/AbstractAccountCredentialCache.java

```diff
@@ -184,6 +188,15 @@ protected List<Credential> getCredentialsFilteredByInternal(@Nullable final Stri
                 matches = matches && StringUtil.equalsIgnoreCaseTrimBoth(clientId, credential.getClientId());
             }
 
+            if (mustMatchOnApplicationIdentifier) {
+                if (credential instanceof AccessTokenRecord) {
+                    final AccessTokenRecord accessToken = (AccessTokenRecord) credential;
+                    matches = matches && StringUtil.equalsIgnoreCaseTrimBoth(applicationIdentifier, accessToken.getApplicationIdentifier());
```

Comment:
if applicationIdentifier is set (and therefore mustMatchOnApplicationIdentifier == true), but accessToken.getApplicationIdentifier() is null, would this fail?

This seems like the upgrade scenario.

Reply:
It's going to be a cache miss... which will result in a new token request.  I didn't think a "migration" solution made sense in this case.  thoughts?

Reply:
If we decide to go that route, I think we should call out that this is sort of a breaking change - so that it doesn't come as a surprise.

(Personally, I don't think doing an additional sign in is a big deal, but I can't speak for others.)

Reply:
I wasn't thinking it would be an additional sign in... as we're talking about access tokens, not refresh tokens at the moment.